osu.Game.Resources/Shaders/sh_ArgonBarPathUtils.h

```diff
@@ -0,0 +1,202 @@
+
+    if (p4.y == p1.y)
+    {
+        return dstToLine(vec2(p1.x + progressRange.x * (p4.x - p1.x), p1.y), vec2(p1.x + progressRange.y * (p4.x - p1.x), p1.y), pixelAbsolutePos);
+    }
+
+    const float curve_start_offset = 70.0;
+    const float curve_end_offset = 40.0;
+    const float curve_smoothness = 10.0;
+
+    mediump float topWidth = max(size.x - pathRadius - curve_start_offset, p1.x) - p1.x;
+    mediump float bottomWidth = p4.x - max(size.x - pathRadius - curve_end_offset, p1.x);
+
+    if (topWidth < bottomWidth)
+    {
+        mediump float newWidth = (topWidth + bottomWidth) * 0.5;
+        topWidth = newWidth;
+        bottomWidth = newWidth;
+    }
+
+    mediump vec2 p2 = vec2(p1.x + topWidth, p1.y);
+    mediump vec2 p3 = vec2(p4.x - bottomWidth, p4.y);
+
+    mediump float slashLength = distance(p2, p3);
+    mediump float roundOffset = min(min(topWidth, slashLength * 0.5), curve_smoothness);
+
+    mediump vec2 arc1Start = vec2(p2.x - roundOffset, p2.y);
+    mediump vec2 arc1End = mix(p2, p3, roundOffset / slashLength);
+    mediump vec2 arc2Start = mix(p2, p3, 1.0 - roundOffset / slashLength);
+    mediump vec2 arc2End = vec2(p3.x + roundOffset, p3.y);
+
+    highp float slashAngle = atan(p2.y - p3.y, p2.x - p3.x) - HALF_PI;
+    slashAngle += float(slashAngle < 0.0) * TWO_PI;
+    slashAngle = TWO_PI - HALF_PI - slashAngle;
+
+    highp float arcAngle = (HALF_PI - slashAngle * 0.5) * 2.0;
+    mediump float arcRadius = roundOffset * tan(slashAngle * 0.5);
+    mediump float arcLength = arcAngle * arcRadius;
+
+    mediump float length1 = arc1Start.x - p1.x;
+    mediump float length2 = distance(arc1End, arc2Start);
+    mediump float length3 = p4.x - arc2End.x;
+    
+    mediump float totalLength = length1 + length2 + length3 + arcLength * 2.0;
+
+    highp float progressArc1Start = length1 / totalLength;
+    highp float progressArc1End = (length1 + arcLength) / totalLength;
+    highp float progressArc2Start = (length1 + arcLength + length2) / totalLength;
+    highp float progressArc2End = (length1 + length2 + arcLength * 2.0) / totalLength;
+    mediump vec2 arc1Centre = vec2(p2.x - roundOffset, p2.y + arcRadius);
+    mediump vec2 arc2Centre = vec2(p3.x + roundOffset, p3.y - arcRadius);
+
+    mediump float minDst = dstToTopLine(progressRange, p1, arc1Start, progressArc1Start, pixelAbsolutePos);
+    minDst = min(minDst, dstToTopArc(progressRange, arcRadius, progressArc1Start, progressArc1End, arc1Centre, slashAngle, arcAngle, pixelAbsolutePos));
+    minDst = min(minDst, dstToSlash(progressRange, arc1End, arc2Start, progressArc1End, progressArc2Start, pixelAbsolutePos));
+    minDst = min(minDst, dstToBottomArc(progressRange, arcRadius, progressArc2Start, progressArc2End, arc2Centre, slashAngle, arcAngle, pixelAbsolutePos));
+    minDst = min(minDst, dstToBottomLine(progressRange, arc2End, p4, progressArc2End, pixelAbsolutePos));
```

Comment:
As I mentioned in [discord](https://discord.com/channels/188630481301012481/188630652340404224/1195596735578394707), when the bar width is set to zero, all of these methods return NaN except for `dstToSlash`. This results in undefined behavior that differs from one platform to another, and needs to be properly guarded against.

Here's a breakdown of all the parameters inserted in these methods when width = 0, to identify the NaNs:

![CleanShot 2024-01-13 at 08 31 46](https://github.com/ppy/osu-resources/assets/22781491/e3408bf4-2f0a-4641-830f-554efeaf00c8)